### DISCUSSION TITLE
Retrieving a RemotePromise out of an Executor callback

### DISCUSSION THREAD

**Discussion by lufriem** [capnproto/capnproto#1981] [Q&A] [answered]
Hello,

I'm still grappling with capnproto so please forgive if my terminology is wrong or confusing..

I'd like to do a client/server RPC application, where the client creates a `capnp::EzRpcClient` in a worker thread, but where I want the main (or any other) thread to access the server's capability. Now I thought the proper way of doing this would be to get `kj::getCurrentThreadExecutor` from the worker thread and use its `executeAsync` (or `executeSync`) functionality to schedule any calls from the main thead to the worker thread.

Invoking the functionality on the server (ie sending the request from the `executeAsync` lambda) seems to work fine, but my problem is that I cannot seem to retrieve the results of my RPC (a `RemotePromise`) back to the calling thread.. if I understand the documentation right, it should work:

>   // The Promise returned by executeAsync() belongs to the requesting thread, not the executor
  // thread. Hence, for example, continuations added to this promise with .then() will execute in
  // the requesting thread.
  //
  // If func() itself returns a Promise, that Promise is *not* returned verbatim to the requesting
  // thread -- after all, Promise objects cannot be used cross-thread. Instead, the executor thread
  // awaits the promise. Once it resolves to a final result, that result is transferred to the
  // requesting thread, resolving the promise that executeAsync() returned earlier.

So I thought the way this would work was simply to `return` the promise (ie the return value of my 'send' call) from func() (my `executeAsync` lambda) but this doesn't compile.. I'm using MSVS2022 (and cpnproto 1.0.4), and I get the error message

`Error	C2440	'<function-style-cast>': cannot convert from 'initializer list' to 'capnp::RemotePromise<>'`

Returning a promise which immediately resolves from func() works just fine, it's just that when I try to return the RemotePromise when things don't work.

Now, I guess I could just forget about doing my `kj::EventLoop` in the worker thread and create a 'one-shot' `capnp::EzRpcClient` for every call (on any thread) but this doesn't seem to be ideal to me (especially as I'd like to keep the worker thread (ie event loop) around to handle server callbacks..)

I'd be most grateful for any pointers how to tackle/fix that! If requested I guess I could also provide some more code..

**Comment by kentonv** [comment]
It's hard to answer this without seeing the actual code. But, making a guess: I suspect the problem is that `executeAsync()` doesn't actually understand that `RemotePromise` is a kind of `Promise`. When you return a Promise from this lambda, the executor waits for the promise to complete (in the executor thread) and only returns the final result to the calling thread. But since it doesn't know that `RemotePromise` is a type of promise, it just tries to return it directly to the calling thread.

To fix this, you need to convert `RemotePromise` to regular `Promise` before returning.

So, instead of:

    return myRpcReq.send();

Try this:

    capnp::RemotePromise<MyResults> rpcPromise = myRpcReq.send();
    kj::Promise<capnp::Response<MyResults>> kjPromise = kj::mv(rpcPromise);
    return kjPromise;

(This works because `capnp::RemotePromise<T>` inherits `kj::Promise<capnp::Response<T>>`.)

**Reply by lufriem** [reply]
Thank you so much for your answer! I'm currently busy with another task but will get back to this next week and keep you updated

**Reply by lufriem** [reply]
First off all, thank you for the answer, this made the compiler happy! I proceeded with development but ran into more issues, I've been trying all kinds of things over the last few days but didn't get anywhere :(

Like mentioned, what I'm trying to do is to:

1. create a worker thread
2. create an `capnp::EzRpcClient` in that thread
3. get my bootstrap capability using that client
4. get the worker thread's `kj::Executor`
5. have the worker thread wait (right now using `kj::NEVER_DONE` but eventually using a `kj::PromiseCrossThreadFulfillerPair<void>`, so that I can shut it down)
6. use the capability's functions from the main thread (or any thread, eventually) using the worker thread's `kj::Executor`

All that works great, the problems start appearing if I try to retrieve the remote function's return value.

I tried using both `kj::Executor::executesync` and `kj::Executor::executeAsync` both fail but (with different symptoms):

`kj::Executor::executeAsync` fails after a single function call (for testing purposes, I call my remote function in a loop). In the lambda this `kj::Exception` is being raised

> Executor's event loop exited before cross-thread event could complete

the worker thread gets the following `kj::Exception` (I guess this actually triggers the failure in the lambda)

> expected threadLocalEventLoop == &loop || threadLocalEventLoop == nullptr; Event armed from different thread than it was created in.  You must use Executor to queue events cross-thread.

If I use `kj::Executor::executesync` to do the cross-thread call, the application simply crashes with the following message

> "null Own<> dereference"

The (quite deep) call stack originates at `>	Calculator_Client.exe!capnp::Response<Calculator::Value::ReadResults>::~Response<Calculator::Value::ReadResults>()	C++
`

**unless** I create a 'kj::EventLoop' in my main thread (which, according to the `kj::Executor::executesync` documentation, should not be required?). In that case, a single function call succeeds, then the second one just 'hangs'

[I forked the 'master' branch of the capnproto repository and stripped down the 'calculator client' sample to reflect what I'm doing](https://github.com/capnproto/capnproto/commit/7311de2f5d90b0d02a309dbdc3e083769bd1ce63). Basically it's just sending a request that evaluates the literal value 123 (ie the first example in the original code), but across threads. I didn't touch the server code, or the interface definition. My 'real' code is actually a bit simpler since I'm not using pipelining, but I get the same errors/behaviour.

Like mentioned, if I do **not** return a promise from the executor lambda, everything works fine but I'd like to have my RPC calls return results so I'm stuck :|

I think the alternative to using `kj::Executor` would be to use a dedicated `capnp::EzRpcClient` instance for each request/thread but since I'd actually like to pass capabilities around this would make things quite awkward..

I'm sure you are super busy but it would be nice to get an opinion if I'm completely off-base with my assumptions on how to do this, or if there's just some error I'm doing, thanks!

**Reply by kentonv** [reply, marked as answer]
Spent a while staring at this, and finally realized what I think is the problem: `capnp::Response` is not safe to pass between threads. A response can potentially contain capabilities, after all, which are tied to the RPC connection they were introduced over (and the EventLoop that owns it). In your use case, the response doesn't actually contain any capabilities, but the implementation of Response is still tied to the RPC connection either way.

I think you will need to clone the response into a thread-safe structure before returning it. A simple way to do that is `capnp::clone(response)`, which will return `Own<ResultType::Reader>`. Assuming the response didn't contain any capabilities, then this value should be safe to pass between threads.

You could also copy the results into some non-capnp structure and return that.

**Reply by lufriem** [reply]
Thanks once more for your time!

So if I have for example a function that expects a password and (upon successful authentication) returns a capabilty, this function **must not** be called using an `kj::Executor`? However, that capability then may then be safely copied to a different thread and calls may be performed using different event loops?

> I think you will need to clone the response into a thread-safe structure before returning it. A simple way to do that is `capnp::clone(response)`, which will return `Own<ResultType::Reader>`.

Sorry, I couldn't get this to work (I think I'm a bit out of my depth here), where would I do that?

> You could also copy the results into some non-capnp structure and return that.

I think I have this (somewhat) figured out now.. I'm no longer returning the `capnp::Response' but a 'simple' Promise, like so (in the 'Calculator' sample)

`// Using the promise, create a pipelined request to call read() on the`
`// returned object, and then send that.`
`auto readPromise = evalPromise.getValue().readRequest().send();`
`return readPromise.then([](capnp::Response<Calculator::Value::ReadResults> &&r) { return r.getValue(); });`

ie the synchronous `kj::Executor` call only returns a double now (the async variant returns a promise for a double).

This approach seems to work just fine in my testing here, I'll play around a bit more with this but things are looking up :) Thank you very much for your help so far! 👍 👍 👍

**Reply by kentonv** [reply]
No, capabilities (`MyInterface::Client` objects) cannot be passed between threads -- they can only be used in the thread that created them. So a capnp message (MessageReader or MessageBuilder) that contains capabilities is not safe to transfer. A message that does not contain any capabilities is safe to transfer. But a capnp::Response (which is a MessageReader plus some other stuff) is not safe. Unfortunately, these rules aren't that well-documented. In general, types associated with Cap'n Proto's serialization layer can be transferred between threads but types associated with RPC or the KJ async promises system cannot (unless otherwise specified).
